src/metrics/include/metrics/v2/event.hpp

```diff
@@ -0,0 +1,282 @@
+public:
+    using std::runtime_error::runtime_error;
+};
+
+
+/**
+ * Manages the collector stub and its interactions.
+ */
+class CollectorStubInterface {
+public:
+    CollectorStubInterface() {
+        setStub();
+    }
+
+    poplar::PoplarEventCollector::StubInterface* operator->() {
+        return _stub.get();
+    }
+
+private:
+    // We should only have one channel across all threads.
+    static std::unique_ptr<poplar::PoplarEventCollector::StubInterface> _stub;
+
+    auto createStub() {
+        grpc::ChannelArguments args;
+        // The BDP estimator overwhelms the server with pings on a heavy workload.
+        args.SetInt(GRPC_ARG_HTTP2_BDP_PROBE, 0);
+        auto channel =
+            grpc::CreateCustomChannel("localhost:2288", grpc::InsecureChannelCredentials(), args);
+        _stub = poplar::PoplarEventCollector::NewStub(channel);
+        return true;
+    }
+
+    void setStub() {
+        static bool stub_created = createStub();
+        return;
+    }
+};
+
+
+/**
+ * Manages the stream of poplar EventMetrics.
+ */
+class StreamInterfaceImpl {
+public:
+    StreamInterfaceImpl()
+        : _options{}, _response{}, _context{}, _stream{_stub->StreamEvents(&_context, &_response)} {
+        _options.set_no_compression().set_buffer_hint();
+    }
+
+    void write(const poplar::EventMetrics& event) {
+        auto success = _stream->Write(event, _options);
+
+        if (!success) {
+            BOOST_THROW_EXCEPTION(PoplarRequestError("Failed to write to stream."));
+        }
+    }
+
+    ~StreamInterfaceImpl() {
+        if (!_stream) {
+            BOOST_LOG_TRIVIAL(error) << "Tried to close grpc stream, but none existed.";
+            return;
+        }
+        if (!_stream->WritesDone()) {
+            BOOST_LOG_TRIVIAL(warning) << "Closing grpc stream, but not all writes completed.";
+        }
+        auto status = _stream->Finish();
+        if (!status.ok()) {
+            BOOST_LOG_TRIVIAL(error)
+                << "Problem closing grpc stream: " << _context.debug_error_string();
+        }
+    }
+
+private:
+    CollectorStubInterface _stub;
+    grpc::WriteOptions _options;
+    poplar::PoplarResponse _response;
+    grpc::ClientContext _context;
+    std::unique_ptr<grpc::ClientWriterInterface<poplar::EventMetrics>> _stream;
+};
+
+/**
+ * Very basic mock for tests.
+ */
+class MockStreamInterface {
+public:
+    MockStreamInterface() {}
+
+    void write(const poplar::EventMetrics& event) {
+        events.push_back(event);
+    }
+
+    // We make this static so we can access it even several private objects deep.
+    static std::vector<poplar::EventMetrics> events;
+};
+
+
+/**
+ * Manages the gRPC-side collector for each operation.
+ * Exists for construction / destruction resource management only.
+ */
+class Collector {
+public:
+    Collector(const Collector&) = delete;
+
+    explicit Collector(const std::string& name, const std::string& path_prefix)
```

Comment:
I'd add a brief comment to these RAII classes that they're only constructed/destructed at setup & teardown time so efficiency isn't a concern and we're really trying to ensure correctness of setup and teardown operations.

Reply:
bump on this